### DISCUSSION TITLE
SetProperty with backing field is missing an overload with callback

### DISCUSSION THREAD

**Discussion by michaldivis** [CommunityToolkit/dotnet#358] [Q&A] [answered]
Hi, I'd like to be able to use this overload of `SetProperty`:

```csharp
bool SetProperty<T>(ref T field, T newValue, Action<T> callback, string? propertyName = null)
```

but that overload doesn't exist. I want to keep using the overload of `SetProperty` that takes a backing field but none of the overloads that take a backing field allow a callback to be passed in as well. Is that intentional?

Thanks!

**Comment by Sergio0694** [comment, marked as answer]
Can you share more about your use case scenario, so we can recommend an alternative? 🙂
That overload is not present for performance reasons, yes, it's by design. What are you trying to do exactly?

**Reply by michaldivis** [reply]
Thanks for the quick answer! The reason I wanted to use the callback is that I'm migrating a codebase that uses async callbacks when a property changed.

Here's an example of what the old code is doing:
```csharp
private Book _selectedBook;
public Book SelectedBook
{
    get => _selectedBook;
    set => SetProperty(ref _selectedBook, value, async book => await LoadBookDetailsAsync(book));
}
```

After migrating to the MVVM toolkit (v8.0.0-preview4), I've switched to invoking the async callback from a synchronous `OnSelectedBookChanged` method (is there an alternative?):
```csharp
[ObservableProperty]
private Book _selectedBook;

partial void OnSelectedBookChanged(Book value)
{
    var taskErrorHandler = DI.Resolve<ITaskErrorHandler>();
    LoadBookDetailsAsync(value).SafeFireAndForget(taskErrorHandler); //the SafeFireAndForget is an async void that catches an exception and forwards it to the ITaskErrorHandler. That's the way I'm currently invoking async code from sync methods
}

private async Task LoadBookDetailsAsync(Book value)
{
    //... reload book details asynchronously
}
```

Is there a better/cleaner way to invoke an asynchronous callback when a property value changes? Thanks for the advice!

**Reply by Sergio0694** [reply]
So, the reason why the overload you're asking for is not there, is that it would cause a closure to be allocated, which hurts performance. You can easily achieve the same functionality with just an explicit branch. That is, you have two ways of doing this: either with a generated property like you're doing, with the logic then put into that `OnPropertyChanged` generated method, or like this:
```csharp
private Book _selectedBook;
public Book SelectedBook
{
    get => _selectedBook;
    set
    {
        if (SetProperty(ref _selectedBook, value))
        {
            var taskErrorHandler = DI.Resolve<ITaskErrorHandler>();
            LoadBookDetailsAsync(value).SafeFireAndForget(taskErrorHandler);
        }
    }
}
```
This is why `SetProperty` returns a `bool`, as it's explicitly meant to enable this scenario 🙂
Would this work for you?

**Reply by michaldivis** [reply]
Yes, this will absolutely work for me. Thank you!
